### DISCUSSION TITLE
Extracting a part of my mesh

### DISCUSSION THREAD

**Discussion by GoldenWitch34** [pyvista/pyvista#2126] [Q&A] [answered]
Hi everyone,

So I'm trying to determine the wall shear stress at the walls of my simulation domain.

To do so, I tried to merge my `internalMesh` with my boundaries to create what I call a `fullMesh`. Here's what I get when I plot the velocity :
<img width="480" alt="Capture_U" src="https://user-images.githubusercontent.com/91612538/152183145-01d2e7b4-93b0-4056-99dd-65d24089dbc4.PNG">

Then, I calculate the velocity gradient using the `compute_derivative` function, here's the data I get from this :
<img width="313" alt="Capture_data" src="https://user-images.githubusercontent.com/91612538/152183871-89c7e27a-c792-44bc-98ff-c4da29b6aef2.PNG">

Now, I'm trying to extract the part of the `fullMesh` where one of my walls is located. I tried doing this with my upper wall, which I refer to as `boundary[2]`.  Here's the data of my `boundary[2]` :
![image](https://user-images.githubusercontent.com/91612538/152186193-9c303e5d-ef6d-430c-aae1-fe717805be4f.png)

I tried using the `clip_surface` function to do so, here's the code I used :
`Diff = deriv.clip_surface(boundary[2], invert=False, value=0.0, compute_distance=False, progress_bar=False)`

Here's the data from `Diff` :
<img width="342" alt="Capture_data_Diff" src="https://user-images.githubusercontent.com/91612538/152187169-9154080a-b841-4b05-b267-e4f24d8a696b.PNG">

Now, the first thing I notice is that the number of cells and points is not the same as that of `boundary[2]`. When I try to plot the velocity at `Diff`, this is what I get :
<img width="469" alt="Capture_U_wall" src="https://user-images.githubusercontent.com/91612538/152187917-02e50577-4540-40cf-b0fa-0dbf12e13027.PNG">

While this is what I get when plotting the velocity at `boundary[2]` : 
<img width="481" alt="Capture_U_wall2" src="https://user-images.githubusercontent.com/91612538/152188282-a0a59838-94cc-48ec-b6ae-7fd3f471dce6.PNG">

So I'm pretty confused, is there a way for me to extract the part of the `fullMesh` that I want? 

Hope this was clear.

**Comment by MatthewFlamm** [comment, marked as answer]
Your boundary mesh points are coincident on points in the internalMesh.  This means that when you merge them, you are likely discarding points from one mesh (the boundary one in this case).  Try changing `main_has_priority` argument in the call to `merge`, if that indeed is what you are doing now.

Another observation is that you are using a turbulent simulation and it looks like you are trying to calculate the derivative of U with respect to the wall.  Maybe for a shear stress calculation.  In turbulent simulations, almost always you cannot do this to get the wall shear stress.  You need to calculate the shear stress using the wall function with the cell centered values of the first off-wall cell (sometimes further off the wall in cases of more exotic wall functions).

**Reply by GoldenWitch34** [reply]
I've made sure that the priority in the `merge` function is set for the boundaries instead of the `internalMesh`, so I don't think that's going to be an issue. 

I see, I didn't take the wall functions into consideration, I thought it would be as easy as just calculating the derivative of U at the wall. 

Thank you for your insight!

**Reply by MatthewFlamm** [reply]
> 've made sure that the priority in the `merge` function is set for the boundaries instead of the `internalMesh`, so I don't think that's going to be an issue.

This sounds like a bug in either pyvista or vtk then.  I get the following incorrect behavior where the vector data `U` is not copied over to the merged mesh.

```python
import pyvista
from pyvista import examples
filename = examples.download_cavity(load=False)
reader = pyvista.get_reader(filename)
reader.set_active_time_value(2.5)
mesh = reader.read()
internal_mesh = mesh['internalMesh']
boundary_mesh = mesh['boundary']['movingWall']
# Using vtk==9.1.0, the point_data is mostly the same on both meshes
# Arbitrarily change the boundary mesh
boundary_mesh.point_data["U"] = 10
merged_mesh = internal_mesh.merge(boundary_mesh, main_has_priority=False)
print(internal_mesh.point_data)
print(boundary_mesh.point_data)
print(merged_mesh.point_data)
```

yields the below

```txt
pyvista DataSetAttributes
Association     : POINT
Active Scalars  : p
Active Vectors  : U
Active Texture  : None
Active Normals  : None
Contains arrays :
    U                       float32  (882, 3)             VECTORS
    p                       float32  (882,)               SCALARS
pyvista DataSetAttributes
Association     : POINT
Active Scalars  : p
Active Vectors  : U
Active Texture  : None
Active Normals  : None
Contains arrays :
    U                       float64  (42,)                VECTORS
    p                       float32  (42,)                SCALARS
pyvista DataSetAttributes
Association     : POINT
Active Scalars  : p
Active Vectors  : None
Active Texture  : None
Active Normals  : None
Contains arrays :
    p                       float32  (882,)               SCALARS
```

**Reply by MatthewFlamm** [reply]
Also, make sure you are using the latest vtk (version 9.1.0).  The reading in of boundary data in OpenFOAM had bugs that were fixed.

**Reply by MatthewFlamm** [reply]
I raise this issue https://github.com/pyvista/pyvista/issues/2156
